src/funnel.ts

```diff
@@ -0,0 +1,280 @@
+ * arguments).
+ *
+ * Typing is inferred from `callback`s param, and from the rest params that
+ * the optional `reducer` function accepts. Use **explicit** types for these
+ * to ensure that everything _else_ is well-typed.
+ *
+ * Notice that this function constructs a funnel **object**, and does **not**
+ * execute anything when called. The returned object should be used to execute
+ * the funnel via the its `call` method.
+ *
+ * Debouncing: use `burstCoolDownMs` and any `invokedAt`.
+ * Throttling: use `delayMs` and `invokedAt: "start"` or `"both"`.
+ * Batching: See the reference implementation in [`funnel.reference-batch.test.ts`](https://github.com/remeda/remeda/blob/main/src/funnel.reference-batch.test.ts).
+ *
+ * @param callback - The main function that would be invoked occasionally based
+ * on `options`. The function would take the latest result of the `reducer`; if
+ * no calls where made since the last time it was invoked it will not be
+ * invoked. If a return value is needed, it should be passed via a reference or
+ * via closure to the outer scope of the funnel.
+ * @param options - An object that defines when `execute` should be
+ * invoked, relative to the calls of `call`. A timer that isn't defined will
+ * **not** be enabled, 0 is not used a default fallback.
+ * @param options.reducer - Combines the arguments passed to `call` with the
+ * value computed on the previous call (or `undefined` on the first time). The
+ * goal of the function is to extract and summarize the data needed for
+ * `callback`. It should be fast and simple as it is called often and should
+ * defer heavy operations to the `execute` function. If the final value
+ * is `undefined`, `callback` will not be called.
+ * @param options.invokedAt - At what "edges" of the funnel's activity
+ * window should `execute` be invoked. `start` means The function will be
+ * invoked immediately (withing the **same** execution frame!), and any
+ * subsequent calls would be ignored until the funnel is idle again. During
+ * this period `reducer` will also not be called. `end` The function will
+ * **not** be invoked initially but the timer will be started. Any calls during
+ * this time would be passed to the reducer, and when the timers are done, the
+ * reduced result would trigger an invocation. When `both` is used The function
+ * will be invoked immediately, and then the funnel would behave as if it was
+ * in the 'end' state. @default 'end'.
+ * @param options.burstCoolDownMs - The burst timer prevents subsequent
+ * calls in short succession to cause excessive invocations (aka "debounce").
+ * This duration represents the **minimum** amount of time that needs to pass
+ * between calls (the "quiet" part) in order for the subsequent call to **not**
+ * be considered part of the burst. In other words, as long as calls are faster
+ * than this, they are considered part of the burst.
+ * @param options.maxBurstDurationMs - Bursts are extended every time a
```

Comment:
**question:** instead of using `maxBurstDurationMs`, what do you think about something like a `maxCallsPerSecond` or `maxCallsPerMinute`?

the thing is, i'm not sure whether thinking in terms of "minimum time between executions" is better than "maximum number of executions per given time period". for me it depends on the use case:

* when i'm, say, typing on a keyboard and fetching with every keystroke, i think more like "i want to call this API a maximum of 2 times per second"
* when i'm doing animation, i think more like "i want to update the screen once every 30ms"

Reply:
That's a great idea also to have those, but they are separate concerns. Debouncing (burst) and Throttling (delay) are timing-based execution limitation mechanisms. 

This is one of the reasons I had a hard time designing this utility. At the end of the day, it makes sense to have a standalone library that provides a whole set of execution-shaping functions so that you can match it to whatever spec you have. If Lodash didn't have debounce/throttle, I would have said this function is out of scope, but as it stands right now these two functions are a blocker for migrating from Lodash for a lot of people.

Reply:
yeah, i see the point now. i really don't think a function like this should belong in remeda, but i see the practicality to including it.